refactor assert_close to be more modular

pmeier opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #67830
* #67796
* #67795
* __->__ #67794

This change is needed to conveniently use the same comparison mechanism for our internal testsuite (see #67796). The reworked version is on par with the previous version except for the ability to pass a custom message as callable. Before we converted everything to a tensor so it was fairly easy to provide consistent mismatch diagnostics to the callable. Now, with arbitrary `Pair`'s that are used for comparison that is no longer viable.

Differential Revision: [D32532206](https://our.internmc.facebook.com/intern/diff/D32532206)

pmeier committed:
refactor assert_close to be more modular

[ghstack-poisoned]

pmeier committed:
Update on "refactor assert_close to be more modular"




[ghstack-poisoned]

pmeier commented on torch/testing/_comparison.py:
```diff
@@ -0,0 +1,939 @@
+    )
+
+
+class UnsupportedInputs(Exception):  # noqa: B903
```
This is a false positive:

```
B903] Data class should either be immutable or use __slots__ to save memory. Use 
collections.namedtuple to generate an immutable class, or enumerate the attributes in a __slot__ 
declaration in the class to leave attributes mutable.
```

See PyCQA/flake8-bugbear#63

pmeier committed:
Update on "refactor assert_close to be more modular"




[ghstack-poisoned]

pmeier committed:
Update on "refactor assert_close to be more modular"




[ghstack-poisoned]

mruberry commented on torch/testing/_comparison.py:
```diff
@@ -0,0 +1,986 @@
+import collections.abc
+from typing import NamedTuple, Callable, Sequence, List, Union, Optional, Type, Tuple, Any, cast
+
+import numpy as np
```
We do require NumPy to run our tests, but torch.testing is a public module that's imported by default:

https://github.com/pytorch/pytorch/blob/191b48b12f33e1e9525882da0c62b68686d69e42/torch/__init__.py#L748

So yes, including NumPy here will make NumPy a dependency of PyTorch. 

We could require NumPy to use torch.testing, but maybe it's easier to just guard the NumPy references here to be conditional on the NumPy module being imported? Creating base classes that are extended if NumPy is available seems like the same solution, so however you want to implement it sounds good.

pmeier replied:
Just to be sure: we do want to have the capability for the `numpy` stuff in the public testing side, i.e. `assert_close`, if they are available, correct? That would extend our current functionality. Note that we are only talking about `numpy` scalars here, which are something different than 0d arrays (I know, weird). `numpy.ndarrays` are supported anyway.

I'm asking, because I could also live with that not being on the public side and only available for internal stuff.

mruberry replied:
It'd be nice to have the NumPy-specific logic in the public torch.testing module if a user is using NumPy. If it's a giant pain to support I could also see doing without.

pmeier committed:
Update on "refactor assert_close to be more modular"




[ghstack-poisoned]

mruberry commented on torch/testing/_comparison.py:
```diff
@@ -0,0 +1,986 @@
+from ._core import _unravel_index
+
+
+class ErrorMeta(NamedTuple):
```
Add a comment/docstring for this class

mruberry commented on torch/testing/_comparison.py:
```diff
@@ -0,0 +1,986 @@
+        return self.type(msg)
+
+
+_DTYPE_PRECISIONS = {
```
Comment for this dict

mruberry commented on torch/testing/_comparison.py:
```diff
@@ -0,0 +1,986 @@
+    def to_error(self) -> Exception:
+        msg = self.msg
+        if self.id:
+            msg += f"\n\nThe failure occurred for item {''.join(str([item]) for item in self.id)}"
```
"item" -> "item(s)"?

pmeier replied:
Depends on what you mean. In `id` we store the nested indices / keys of the error. This does not multiple items, but rather a nested one. Example:

```python
>>> actual = dict(foo=[0, dict(bar=1, baz=[2])])
>>> expected = dict(foo=[0, dict(bar=1, baz=[3])])
>>> torch.testing.assert_close(actual, expected)
[...]
The failure occurred for item ['foo'][1]['baz'][0]
```

As long as we are talking about builtin types as keys one could now use the message to extract the values that are causing the error:
```python
>>> actual['foo'][1]['baz'][0]
2
>>> expected['foo'][1]['baz'][0]
3
```

mruberry replied:
Ah, thanks for the clarification

mruberry commented on torch/testing/_comparison.py:
```diff
@@ -0,0 +1,986 @@
+}
+
+
+def default_tolerances(*inputs: Union[torch.Tensor, torch.dtype]) -> Tuple[float, float]:
```
Comment/docstring for this function

mruberry commented on torch/testing/_comparison.py:
```diff
@@ -0,0 +1,986 @@
+    return max(rtols), max(atols)
+
+
+def parse_tolerances(
```
Comment or docstring

mruberry commented on torch/testing/_comparison.py:
```diff
@@ -0,0 +1,986 @@
+        return None, default_tolerances(*inputs)
+
+
+def _make_mismatch_msg(
```
Comment/docstring

mruberry commented on torch/testing/_comparison.py:
```diff
@@ -0,0 +1,986 @@
+    return msg
+
+
+def make_scalar_mismatch_msg(
```
(let's pretend I mentioned adding a comment or docstring to every function)

mruberry reviewed (commented):
Overall this seems reasonable, we just need to fix the NumPy import issue. I made a few inline comments suggesting we improve code readability with more comments and docstrings. An overview of the approach in a comment would be nice, too, so future maintainers can follow along.

pmeier committed:
Update on "refactor assert_close to be more modular"




[ghstack-poisoned]

pmeier commented:
@mruberry I've added `numpy` as optional dependency. Plus. I've added a bunch of docstrings. LMK if you feel like I missed something important.

pmeier committed:
Update on "refactor assert_close to be more modular"


This change is needed to conveniently use the same comparison mechanism for our internal testsuite (see #67796). The reworked version is on par with the previous version except for the ability to pass a custom message as callable. Before we converted everything to a tensor so it was fairly easy to provide consistent mismatch diagnostics to the callable. Now, with arbitrary `Pair`'s that are used for comparison that is no longer viable.

[ghstack-poisoned]

pmeier committed:
Update on "refactor assert_close to be more modular"


This change is needed to conveniently use the same comparison mechanism for our internal testsuite (see #67796). The reworked version is on par with the previous version except for the ability to pass a custom message as callable. Before we converted everything to a tensor so it was fairly easy to provide consistent mismatch diagnostics to the callable. Now, with arbitrary `Pair`'s that are used for comparison that is no longer viable.

[ghstack-poisoned]

mruberry commented on torch/testing/_comparison.py:
```diff
@@ -0,0 +1,1141 @@
+    Returns:
+        (List[Pair]): Originated pairs.
+    """
+    # We explicitly exclude str's here since they are self-referential and would cause an infinite recursion loop:
```
Nice

mruberry reviewed (approved):
It's really cool to see how much knowledge and thoughtfulness has culminated in this version of allclose

pmeier committed:
Update on "refactor assert_close to be more modular"


This change is needed to conveniently use the same comparison mechanism for our internal testsuite (see #67796). The reworked version is on par with the previous version except for the ability to pass a custom message as callable. Before we converted everything to a tensor so it was fairly easy to provide consistent mismatch diagnostics to the callable. Now, with arbitrary `Pair`'s that are used for comparison that is no longer viable.

[ghstack-poisoned]

mruberry commented:
@mruberry has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D32532206).
